proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+
+## Null tracking for variables
+
+For certain expressions denoting variables or properties, the null state is tracked between occurrences, based on assignments to them, tests performed on them and the control flow between them. This is similar to how definite assignment is tracked for variables. The tracked expressions are the ones of the following form:
+
+```antlr
+tracked_expression
+    : simple_name
+    | this
+    | base
+    | tracked_expression '.' identifier
+    ;
+```
+
+Where the identifiers denote fields or properties.
+
+***Describe null state transitions similar to definite assignment***
+
+## Null state for expressions
+
+The null state of an expression is derived from its form and type, and from the null state of variables involved in it.
+
+## Literals
+
+The null state of a `null` literal is "maybe null". The null state of a `default` literal that is being converted to a type that is known not to be a nonnullable value type is "maybe null". The null state of any other literal is "not null".
+
+## Simple names
+
+If a `simple_name` is not classified as a value, its null state is "not null". Otherwise it is a tracked expression, and its null state is its tracked null state at this source location.
+
+## Member access
+
+If a `member_access` is not classified as a value, its null state is "not null". Otherwise, if it is a tracked expression, its null state is its tracked null state at this source location. Otherwise, if the nullability of its type is nullable or unknown, the null state is "maybe null". Otherwise the null state is "not null".
+
+## Invocation expressions
+
+If an `invocation_expression` invokes a member that is declared with one or more attributes for special null behavior, the null state is determined by those attributes. Otherwise, if the nullability of the type of the expression is nullable or unknown, the null state is "maybe null". Otherwise the null state is "not null".
+
+## Element access
+
+If an `element_access` invokes an indexer that is declared with one or more attributes for special null behavior, the null state is determined by those attributes. Otherwise, if the nullability of the type of the expression is nullable or unknown, the null state is "maybe null". Otherwise the null state is "not null".
+
+## Base access
+
+If `B` denotes the base type of the enclosing type, `base.I` has the same null state as `((B)this).I` and `base[E]` has the same null state as `((B)this)[E]`.
+
+## Default expressions
+
+`default(T)` has the null state "non-null" if `T` is known to be a nonnullable value type. Otherwise it has the null state "maybe null".
+
+## Null-conditional expressions
+
+A `null_conditional_expression` has the null state "maybe null".
```

Comment:
is this appropriate if the 'null state' of the expression in `expr?.` is non-null?  I guess it is since the 'non-null' state we have may be incorrect, and this is a pretty explicit way for the user to be explicit that they can handle null and will get null back in that case.